### PR TITLE
add dense/sparse biencoder models and refactor modeling module

### PR DESCRIPTION
**Pull request by MXueguang**

### CODE REVIEW

File: examples/splade/train_splade.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+class SpladeTrainingArguments(TevatronTrainingArguments):
+    q_flops_loss_factor: float = field(default=4)
+    p_flops_loss_factor: float = field(default=32)
+
+
+class SpladeTrainer(EncoderTrainer):
+    def __init__(self, *args, **kwargs):
+        super(SpladeTrainer, self).__init__(*args, **kwargs)
+
+    def _flops(self, inputs):
```
Comment (luyug) [inline examples/splade/train_splade.py:31]: make this a static method or maybe pull it into loss definition file?

File: examples/unicoil/README.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
+## Train UniCOIL on MS MARCO
+```bash
+CUDA_VISIBLE_DEVICES=0 python examples/unicoil/train_unicoil.py \
```
Comment (luyug) [inline examples/unicoil/README.md:3]: should we include the general coil setup and make unicoil as a specific setup flag

File: src/tevatron/trainer.py [hunk 3]
```diff
@@ -20,9 +20,9 @@
     _grad_cache_available = False
 
 
-class DenseTrainer(Trainer):
+class EncoderTrainer(Trainer):
```
Comment (luyug) [inline src/tevatron/trainer.py:23]: Now I think about it, EncoderTrainer still feels a little bit weird. Can we just call this Trainer? If not, maybe Tevatron trainer?

**Comment by cadurosar** [comment]
Just to say thanks for sharing your very efficient framework and that I was able to test SPLADE with this PR implementation. 
I only had to add a pyserini encoder (for eval) / change the Hard Negative code to use pyserini output (can add a PR later if you think it suits the project). Other than that the process was pretty smooth and results were pretty good.

File: examples/unicoil/README.md (hunk 2 above)
Comment (MXueguang) [reply examples/unicoil/README.md:3]: yeah, let's do it in a separate PR?

**Comment by MXueguang** [comment]
Hi @cadurosar 
Thank you very much for testing the SPLADE implementation. 
And we look forward to your PR.

File: examples/unicoil/README.md (hunk 2 above)
Comment (luyug) [reply examples/unicoil/README.md:3]: if we consider unicoil as a special case of coil, it may make sense to keep a single coil class for both. in that case, we probably needs to delay unicoil to the next pr. 
otherwise, we ship them as two separate classes, one in this one and one in the future pr.
